src/pixel/heterogeneous_pixel.rs

```diff
@@ -0,0 +1,217 @@
+    type SelfType<U: PixelComponent, V: PixelComponent>: HeterogeneousPixel<
+        SelfType<Self::ColorComponent, Self::AlphaComponent> = Self,
+    >;
+    //TODO if const generic expressions become stable remove this associated type and just use
+    //`[Self::Component; Self::COMPONENT_COUNT]` as the return type.
+    //
+    /// The color array form of `Self`
+    type ColorArray<R>: IntoIterator<Item = R> + AsSlice<R>;
+
+    /// Converts an owned `Pixel` type to an array of its color components.
+    fn color_array(&self) -> Self::ColorArray<Self::ColorComponent>;
+    /// Returns the alpha component of the pixel if it has one.
+    fn alpha(&self) -> Option<Self::AlphaComponent>;
+
+    /// Create a new instance given an array of its color components and an alpha component.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the colors iterator does not produce enough color components for the pixel.
+    fn from_colors_alpha(
+        colors: impl IntoIterator<Item = Self::ColorComponent>,
+        alpha: Self::AlphaComponent,
+    ) -> Self;
+
+    /// Maps each of the pixels color components with a function `f` to any other type.
+    ///
+    /// See [`HeterogeneousPixel::map_colors_same()`] if you want to map the color components to the
+    /// same type.
+    fn map_colors<U>(
+        &self,
+        f: impl FnMut(Self::ColorComponent) -> U,
+    ) -> Self::SelfType<U, Self::AlphaComponent>
+    where
+        U: PixelComponent;
+    /// Maps each of the pixels color components with a function `f` to the same type.
+    ///
+    /// See [`HeterogeneousPixel::map_colors()`] if you want to map the color components to a
+    /// different type.
+    fn map_colors_same(&self, f: impl FnMut(Self::ColorComponent) -> Self::ColorComponent) -> Self;
+
+    /// Maps the pixels alpha component with a function `f` to any other type.
+    ///
+    /// If the pixel has no alpha component then the pixel is returned unchanged.
+    ///
+    /// See [`HeterogeneousPixel::map_alpha_same()`] if you want to map the alpha component to the
+    /// same type.
+    fn map_alpha<U>(
+        &self,
+        f: impl FnMut(Self::AlphaComponent) -> U,
+    ) -> Self::SelfType<Self::ColorComponent, U>
+    where
+        U: PixelComponent;
+    /// Maps the pixels alpha component with a function `f` to the same type.
+    ///
+    /// If the pixel has no alpha component then the pixel is returned unchanged.
+    ///
+    /// See [`HeterogeneousPixel::map_alpha()`] if you want to map the alpha component to a
+    /// different type.
+    fn map_alpha_same(&self, f: impl FnMut(Self::AlphaComponent) -> Self::AlphaComponent) -> Self;
+}
+
+macro_rules! without_alpha {
+    ($name:ident, $length:literal, [$($color_bit:tt),*]) => {
+        impl<T> HeterogeneousPixel for $name<T>
+        where
+            T: PixelComponent,
+        {
+            type ColorComponent = T;
+            type AlphaComponent = T;
```

Comment:
Why not `()`? Having working `map_alpha` on non-alpha color formats is unfortunate.

Reply:
Good point!

Reply:
Actually but then what happens to the relationship between the two traits:
```rust
pub trait HomogeneousPixel:
    HeterogeneousPixel<ColorComponent = Self::Component, AlphaComponent = Self::Component>
```

Reply:
We could remove the super-trait relationship but then a user would have to specify both pixel bounds to use functions from either method as opposed to just specifying `HomogeneousPixel` and then `HeterogeneousPixel` methods come for free.

Reply:
The bound `(): PixelComponent` also makes this harder to use in generality.

Reply:
That's a bummer. Maybe the trait should keep all alpha methods fallible (`alpha_opt`/`try_alpha`), and put non-fallible methods as either inherent methods on types, or on some `PixelAlpha` trait?

Reply:
I think the whole point of the `HeterogeneousPixel` type is to be able to write functions on types that may/maynot contain an alpha component, this inherently requires implementations for pixel types which don't have an alpha component to have no-ops. Like:
```rust
impl HeterogeneousPixel for Rgb {
    fn alpha_checked(&self) -> Option<T> {
        None
    }
}
```

I've added the `HasAlpha` trait for scenarios where people want to implement function for pixels which **must** have alpha components which can have methods like `fn alpha(&self) -> T`.